### DISCUSSION TITLE
Optimizing C array lookups for the 6502 — 8bitworkshop documentation

### DISCUSSION THREAD

**Discussion by giscus** [sehugg/8bitworkshop#81] [General]
# Optimizing C array lookups for the 6502 — 8bitworkshop documentation

Most of the platforms in the IDE that are powerful enough to support a C compiler are Z80-based. The Z80 isn’t the easiest target for C, but at least it has a lot of registers. While the 6502 only ...

https://8bitworkshop.com/docs/posts/2019/cc65-optimization.html

**Comment by HenrikSandell** [comment]
I have spent some time writing a C compiler for 6502. I have made a couple of different design decisions compared to cc65.

Below is the code my compiler generated for the testprogram; I have added some comments.
(Please note that I did not optimise my compiler for this problem; although I'm inspired to do so...).

    ; int getValue(int index) {
    getValue:
    ;     return (arr16[index & 0x7f] & 0xff) >> 1;
    LDX $1f.    # I use $1f as kind of a BP into my auto/data stack located between $20-$f0 in ZeroPage
    LDA  $1E,X  # Load index low part, in 8086 this would be MOV AL,-2[BP]
    AND #127 
    STA $A4+0   # Store in temp variable
    LDA #0      # High word will be zero since we do an and with an int with less than 256 
    STA $A4+1. 
    LDA $A4+0   # Shift with 1 to get array adress since each element is two bytes
    ASL A
    STA $A6+0
    LDA $A4+1. 
    ROL A     # Use ROL since we need to include carry
    STA $A6+1
    LDA $A6+0  # Add offset to array; in this case the array is put in adress $200; just copy low byte
    CLC
    STA $A2+0
    LDA $A6+1  # Add two to high byte
    ADC #2
    STA $A2+1  # Store in $A2 and $A3
    LDY #0     # Read from ($A2),Y; thus we need Y to be zero
    LDA  ($A2),Y
    STA $A0+0  # Store lower part of word in temp
    TYA        # High word will be zero since we take & 0xff. 
               # My peephole optimizer luckily find a 0 in Y; thus I use TYA instead of LDA #0
    STA $A0+1  # Store in temp (Would be even better to do STY $A0+1)
    LSR $A0+1  # Shift right; we can do this in place
    ROR $A0+0  # Rotate since we need to have lower bit from high byte
    LDA $A0+0  # I return in ($FC,$FD) by convention; transfer ($A0,$A1) to ($FC,$FD)
    STA $FC+0  # This bugs me; I should have been able to use $FC instead of $A0...
    LDA $A0+1
    STA $FC+1
    RTS
